persistence/sql/persister_registration.go

```diff
@@ -64,3 +70,133 @@ func (p *Persister) DeleteExpiredRegistrationFlows(ctx context.Context, expiresA
+	if err := sqlcon.HandleError(p.GetConnection(ctx).Transaction(func(tx *pop.Connection) error {
+		if err := tx.RawQuery(fmt.Sprintf("UPDATE %s SET submit_count = submit_count + 1 WHERE id = ? AND nid = ?", flowTableName), flowID, nid).Exec(); err != nil {
+			return err
+		}
+
+		var submitCount int
+		// Because MySQL does not support "RETURNING" clauses, but we need the updated `submit_count` later on.
+		//#nosec G201 -- TableName is static
+		if err := sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("SELECT submit_count FROM %s WHERE id = ? AND nid = ?", flowTableName), flowID, nid).First(&submitCount)); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+			return err
+		}
+
+		// This check prevents parallel brute force attacks to generate the recovery code
+		// by checking the submit count inside this database transaction.
+		// If the flow has been submitted more than 5 times, the transaction is aborted (regardless of whether the code was correct or not)
+		// and we thus give no indication whether the supplied code was correct or not. See also https://github.com/ory/kratos/pull/2645#discussion_r984732899
+		if submitCount > 5 {
+			return errors.WithStack(code.ErrCodeSubmittedTooOften)
+		}
+
+		var registrationCodes []code.RegistrationCode
+		if err := sqlcon.HandleError(tx.Where("nid = ? AND selfservice_registration_flow_id = ?", nid, flowID).All(&registrationCodes)); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+
+			return err
+		}
+
+	secrets:
+		for _, secret := range p.r.Config().SecretsSession(ctx) {
+			suppliedCode := []byte(p.hmacValueWithSecret(ctx, rawCode, secret))
+			for i := range registrationCodes {
+				code := registrationCodes[i]
+				if subtle.ConstantTimeCompare([]byte(code.CodeHMAC), suppliedCode) == 0 {
+					// Not the supplied code
+					continue
+				}
+				registrationCode = &code
+				break secrets
+			}
+		}
+
+		if registrationCode == nil || !registrationCode.IsValid() {
+			// Return no error, as that would roll back the transaction
+			return nil
+		}
+
+		//#nosec G201 -- TableName is static
+		return sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("UPDATE %s SET used_at = ? WHERE id = ? AND nid = ?", registrationCode.TableName(ctx)), time.Now().UTC(), registrationCode.ID, nid).Exec())
+	})); err != nil {
+		return nil, err
+	}
+
+	if registrationCode == nil {
+		return nil, errors.WithStack(code.ErrCodeNotFound)
+	}
+
+	if registrationCode.IsExpired() {
+		return nil, errors.WithStack(flow.NewFlowExpiredError(registrationCode.ExpiresAt))
+	}
+
+	if registrationCode.WasUsed() {
+		return nil, errors.WithStack(code.ErrCodeAlreadyUsed)
+	}
+
+	// ensure that the identifiers extracted from the traits are contained in the registration code
+	if !slice.Contains(addresses, registrationCode.Address) {
```

Comment:
Can it happen that `addresses` has `eXaMpLe@uppercase.com` and `code.Address` `example@uppercase.com` where the email is the same but we end up with an error here? Aka missing normalization?

Reply:
AFAIK the identity validation step normalizes the credentials which is reused as the address at a later point. So the stored address would be a normalized version of the identifier submitted by the user. 

We could enforce a normalization strategy here just to be sure.